internal/handlers/common/findandmodify.go

```diff
@@ -168,3 +178,152 @@ func GetFindAndModifyParams(doc *types.Document, l *zap.Logger) (*FindAndModifyP
 		MaxTimeMS:          maxTimeMS,
 	}, nil
 }
+
+// UpsertDocument updates the first document if exists, or create an insert document
+// from the params if no documents in query result docs.
+func UpsertDocument(docs []*types.Document, params *FindAndModifyParams) (*UpsertParams, error) {
+	res := new(UpsertParams)
+	var err error
+
+	if len(docs) == 0 {
+		res.Insert, err = insertDocuments(params)
+
+		res.ReturnValue = types.Null
+		if params.ReturnNewDocument {
+			res.ReturnValue = res.Insert
+		}
+
+		return res, err
+	}
+
+	res.Update, err = updateDocuments(docs, params)
+
+	res.ReturnValue = docs[0]
+	if params.ReturnNewDocument {
+		res.ReturnValue = res.Update
+	}
+
+	return res, err
+}
+
+// insertDocuments creates an insert document from the parameter.
+// When inserting new document we must check that `_id` is present, so we must extract `_id`
+// from query or generate a new one.
+func insertDocuments(params *FindAndModifyParams) (*types.Document, error) {
+	insert := must.NotFail(types.NewDocument())
+
+	if params.HasUpdateOperators {
+		if _, err := UpdateDocument(insert, params.Update); err != nil {
+			return nil, err
+		}
+	} else {
+		insert = params.Update
+	}
+
+	if !insert.Has("_id") {
+		id, err := getUpsertID(params.Query)
+		if err != nil {
+			return nil, err
+		}
+
+		insert.Set("_id", id)
+	}
+
+	return insert, nil
+}
+
+// updateDocuments updates the first document with update parameters.
+func updateDocuments(docs []*types.Document, params *FindAndModifyParams) (*types.Document, error) {
+	update := docs[0].DeepCopy()
+
+	if params.HasUpdateOperators {
+		if _, err := UpdateDocument(update, params.Update); err != nil {
+			return nil, err
+		}
+
+		return update, nil
+	}
+
+	for _, k := range params.Update.Keys() {
+		v := must.NotFail(params.Update.Get(k))
+		if k == "_id" {
+			return nil, commonerrors.NewCommandError(
+				commonerrors.ErrImmutableField,
+				fmt.Errorf(
+					"Plan executor error during findAndModify :: caused "+
+						"by :: After applying the update, the (immutable) field "+
+						"'_id' was found to have been altered to _id: \"%s\"",
+					v,
+				),
+			)
+		}
+
+		update.Set(k, v)
+	}
+
+	return update, nil
+}
+
+// getUpsertID gets the _id to use for upsert document. If query contains _id,
+// that _id is assigned unless _id contains operator. Otherwise, it generates an ID.
+func getUpsertID(query *types.Document) (any, error) {
+	id, err := query.Get("_id")
+	if err != nil {
+		return types.NewObjectID(), nil
+	}
+
+	idDoc, ok := id.(*types.Document)
+	if !ok {
+		return id, nil
+	}
+
+	_, hasOp, err := hasFilterOperator(idDoc)
+	if err != nil {
+		return nil, err
+	}
+
+	if hasOp {
+		return types.NewObjectID(), nil
+	}
```

Comment:
This is an interesting piece of logic. Do we have a compat test that covers it?

Reply:
Yes `UpsertExpressionKey`. And this was the core cause of https://github.com/FerretDB/FerretDB/issues/2241